### PR TITLE
Rework tokenization and checking pipeline

### PR DESCRIPTION
**Pull request by craigp-atw**
Implement parser, update checking pipeline, and update unit tests.

All the unit tests currently pass.

Before: random test text
image

After: random test text. fixed certain punctuation issues, including allowing words like .1
image

### CODE REVIEW

File: app/src/lib/index.js
Comment (craigp-atw) [inline app/src/lib/index.js:1]: Moved function_words.js because it's now only used in the parser

**Review by longrunningprocess — CHANGES_REQUESTED** [review]
some MAJOR improvements in here my man, I'm really impressed!

File: app/src/lib/parser/tokenize.js [hunk 1]
```diff
@@ -1,73 +1,304 @@
...
+					// simple/
+					eat_until(REGEXES.TOKEN_END_BOUNDARY)
+					return create_error_token('Pairings should have the form simple/complex, e.g., follower/disciple.')
+				}
+			} else if (peek_match_two(/\.\d/)) {
+				// may be a decimal number like 2.5
+				advance()	// consume the '.'
+				eat(/\d/)
+				return check_boundary_for_token(TOKEN_TYPE.WORD)
+			} else {
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/parser/tokenize.js:55]: For a simple if/else I'm fine removing the else but for longer or more complicated else-if chains I think it's beneficial to have the else to clearly identify it as part of the chain.

File: app/src/lib/parser/tokenize.js [hunk 2]
```diff
@@ -1,73 +1,304 @@
...
+				// (imp
+				return create_error_token('Missing a closing parenthesis')
+			} else if (!is_at_end() && !peek_match(REGEXES.ANY_WHITESPACE)) {
+				// (imp)[ (imp)token
+				eat_until(REGEXES.TOKEN_END_BOUNDARY)
+				return create_error_token('Must include a space after a clause notation')
+			} else if (!CLAUSE_NOTATIONS.includes(collect_text())) {
+				// (invalid-notation)
+				return create_error_token('This clause notation is not recognized')
+			} else {
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/parser/tokenize.js:74]: For a simple if/else I'm fine removing the else but for longer or more complicated else-if chains I think it's beneficial to have the else to clearly identify it as part of the chain. And then it's clear that each branch is responsible for returning.

File: app/src/lib/parser/tokenize.js [hunk 3]
```diff
@@ -1,73 +1,304 @@
...
+
+		} else if (match_char('.')) {
+			if (peek_match(REGEXES.TOKEN_END_BOUNDARY)) {
+				// regular punctuation use
+				return create_simple_token(TOKEN_TYPE.PUNCTUATION)
+			} else if (match_regex(/\d/)) {
+				// decimal number like .1
+				eat(/\d/)
+				return check_boundary_for_token(TOKEN_TYPE.WORD)
+			} else {
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/parser/tokenize.js:94]: To add to my above comments, I see the return statement outside of the if-else chain as the 'default' or 'valid' or 'proper' value and anything in the chain as a guard clause of some sort. I don't feel that is the case in this branch, so I would like to keep the else if that's ok.

File: app/src/lib/regexes.js [hunk 4]
```diff
@@ -9,60 +9,43 @@
...
-//			'/': alternate forms
-//			'(' or ')': clause notations or pronoun referents
-//			"'": possessives, .e.g, Israelites'
-//			'-': hyphentated clause notations
-// (\W*)$: group 3 => same as group 1 but at the end of the string (if they appear)
-const PUNCTUATION_GROUPING = /^(\W*?)([\w/()'-]*)(\W*)$/
-
-const POSSESSIVE = /'s?/
+const TOKEN_END_BOUNDARY = /[\s.,!?:"\]]/
+const WORD_START_CHAR = /[a-zA-Z0-9-]/
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/regexes.js:25]: yes, the underscore

File: app/src/lib/regexes.js [hunk 5]
```diff
@@ -9,60 +9,43 @@
...
-const IS_PAIRING = /^\w+\/\w+$/
 const IS_LEVEL_SIMPLE = /[01]/
 const IS_LEVEL_COMPLEX = /[23]/
 
+/**
+ * @param {RegExp} regex1
+ * @param {RegExp} regex2
+ */
+const OR = function (regex1, regex2) {
+	return new RegExp(regex1.source + '|' + regex2.source)
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/regexes.js:36]: not at all!

File: app/src/lib/token.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+export const TOKEN_TYPE = {
+    ERROR,
+    PUNCTUATION,
+    NOTE,
+    FUNCTION_WORD,
+    WORD,
+    PAIRING,
+}
+
+export const DEFAULT_TOKEN_VALUES = {
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/token.js:17]: It is not a full Token. I purposefully avoided including 'token' and 'type' so you're forced to include those when creating a token. This should only be used for filling in the irrelevant fields for a particular token type using the spread operator.

File: app/src/lib/parser/tokenize.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply app/src/lib/parser/tokenize.js:55]: I understand this approach better now, thanks

File: app/src/lib/parser/tokenize.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply app/src/lib/parser/tokenize.js:74]: I understand this approach better now, thanks

File: app/src/lib/parser/tokenize.js
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/parser/tokenize.js:1]: Thanks! I learned most of what I know by going through https://craftinginterpreters.com/

File: app/src/lib/parser/tokenize.js [hunk 7]
```diff
@@ -1,73 +1,304 @@
...
+	 */
+	function peek_match(regex) {
+		return !is_at_end() && peek().match(regex)
+	}
 
-			return tokens
+	/**
+	 * @param {RegExp} regex
+	 */
+	function match_regex(regex) {
```
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/parser/tokenize.js]: Removed match_char() and renamed match_regex() to just match()

File: app/src/lib/token.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (craigp-atw) [reply app/src/lib/token.js]: Reworked to use a create_token() function with default arguments

File: app/src/lib/parser/function_words.js [hunk 8]
```diff
@@ -13,10 +15,12 @@ export const function_words = [
 	'of',
 	'probably',
 	'should',
+	'start',
+	'started',
+	'starts',
 	'stop',
 	'stopped',
 	'stops',
-	'to',
```
Comment (craigp-atw) [inline app/src/lib/parser/function_words.js:19]: there were two 'to's

File: app/src/lib/parser/function_words.js [hunk 9]
```diff
@@ -1,8 +1,10 @@
 // prettier-ignore
-export const function_words = [
+export const FUNCTION_WORDS = new Set([
 	'a',
 	'an',
 	'certainly',
+	'continue',
```
Comment (craigp-atw) [inline app/src/lib/parser/function_words.js:6]: While 'continue' is in the ontology, it is only used to determine the verb aspect within the analyzer, never as a source concept.

File: app/src/lib/parser/function_words.js [hunk 10]
```diff
@@ -13,10 +15,12 @@ export const function_words = [
 	'of',
 	'probably',
 	'should',
+	'start',
```
Comment (craigp-atw) [inline app/src/lib/parser/function_words.js:18]: There is a word 'start' in the ontology, but the way it is written in the phase 1 is different. 'start' is always used in the analyzer to determine the inceptive aspect of a verb and can be treated as a function word.

File: app/src/lib/parser/tokenize.js [hunk 11]
```diff
@@ -1,73 +1,312 @@
...
 /**
  * @param {string} text
  * @returns {Token[]}
  */
 export function tokenize_input(text = '') {
-	return text
-		.split(REGEXES.ANY_WHITESPACE)
-		.filter(not_empty)
-		.map(token => token)
+	let tokens = []
```
Comment (longrunningprocess) [inline app/src/lib/parser/tokenize.js:12]:
```suggestion
	const tokens = []
```

File: app/src/lib/parser/tokenize.js
Comment (longrunningprocess) [inline app/src/lib/parser/tokenize.js:1]: already much easier to read, thank you!

File: app/src/lib/tokens/FunctionWord.svelte
Comment (longrunningprocess) [inline app/src/lib/tokens/FunctionWord.svelte:1]: better with the generic name here 🏆

File: app/src/lib/tokens/Pairing.svelte [hunk 12]
```diff
@@ -3,38 +3,34 @@
...
 	import {REGEXES} from '$lib/regexes'
+	import {create_error_token} from '$lib/parser/token'
 
-	/** @type {CheckedToken} */
-	export let checked_token
+	/** @type {Token} */
+	export let token
 
-	$: [left, right] = split_pair(checked_token.token)
+	// At this point the pairing tokens will never be null. This satisfies the compiler
```
Comment (longrunningprocess) [inline app/src/lib/tokens/Pairing.svelte:14]: maybe a `// ts-ignore` instead of the extra code?

**Review by longrunningprocess — CHANGES_REQUESTED** [review]
everything is soooo much cleaner!

File: app/src/lib/tokens/Pairing.svelte (hunk 12 above)
Comment (craigp-atw) [reply app/src/lib/tokens/Pairing.svelte:14]: Good call

**Review by longrunningprocess — APPROVED** [review]
masterpiece!

**Pull request merged**